pkg/genericapiserver/server_run_options.go

```diff
@@ -55,28 +64,104 @@ type ServerRunOptions struct {
 	SecurePort                int
 	ServiceClusterIPRange     net.IPNet // TODO: make this a list
 	ServiceNodePortRange      utilnet.PortRange
-	TLSCertFile               string
-	TLSPrivateKeyFile         string
+	StorageVersions           string
+	// The default values for StorageVersions. StorageVersions overrides
+	// these; you can change this if you want to change the defaults (e.g.,
+	// for testing). This is not actually exposed as a flag.
+	DefaultStorageVersions string
+	TLSCertFile            string
+	TLSPrivateKeyFile      string
 }
 
 func NewServerRunOptions() *ServerRunOptions {
 	return &ServerRunOptions{
-		APIGroupPrefix:       "/apis",
-		APIPrefix:            "/api",
-		BindAddress:          net.ParseIP("0.0.0.0"),
-		CertDirectory:        "/var/run/kubernetes",
-		EnableLogsSupport:    true,
-		EnableProfiling:      true,
-		EnableWatchCache:     true,
-		InsecureBindAddress:  net.ParseIP("127.0.0.1"),
-		InsecurePort:         8080,
-		LongRunningRequestRE: defaultLongRunningRequestRE,
-		MasterCount:          1,
-		MaxRequestsInFlight:  400,
-		MinRequestTimeout:    1800,
-		RuntimeConfig:        make(config.ConfigurationMap),
-		SecurePort:           6443,
+		APIGroupPrefix:         "/apis",
+		APIPrefix:              "/api",
+		BindAddress:            net.ParseIP("0.0.0.0"),
+		CertDirectory:          "/var/run/kubernetes",
+		EnableLogsSupport:      true,
+		EnableProfiling:        true,
+		EnableWatchCache:       true,
+		InsecureBindAddress:    net.ParseIP("127.0.0.1"),
+		InsecurePort:           8080,
+		LongRunningRequestRE:   defaultLongRunningRequestRE,
+		MasterCount:            1,
+		MaxRequestsInFlight:    400,
+		MinRequestTimeout:      1800,
+		RuntimeConfig:          make(config.ConfigurationMap),
+		SecurePort:             6443,
+		StorageVersions:        registered.AllPreferredGroupVersions(),
+		DefaultStorageVersions: registered.AllPreferredGroupVersions(),
+	}
+}
+
+// StorageGroupsToEncodingVersion returns a map from group name to group version,
+// computed from the s.DeprecatedStorageVersion and s.StorageVersions flags.
+func (s *ServerRunOptions) StorageGroupsToEncodingVersion() (map[string]unversioned.GroupVersion, error) {
```

Comment:
Yes this is to build a ResourceEncodingConfig.
I assume you mean that this should return DefaultResourceEncodingConfig.

I think DefaultResourceEncodingConfig is a lot more powerful than what this function returns. For ex: this only has StorageEncoding and no InMemoryEncoding. Also this does not have per resource overrides.
If I change this to return DefaultResourceEncodingConfig, then I will also have to add more logic to the code in StorageFactoryBuilder to merge two DefaultResourceEncodingConfigs.
I think we will be adding some unnecessary complexity by adding that.

I think it will be good to make that change, when in future, we allow users to do more of what DefaultResourceEncodingConfig allows (example specify per resource overrides in command line arguments)